File: solutions/java/circular-buffer/1/src/main/java/CircularBuffer.java
class CircularBuffer<T> {

    Node<T> tail, head;

    CircularBuffer(final int size) {
        Node<T> current = new Node<>(null, null);
        head = current;
        for (int i = 1; i < size; i++) {
            current = new Node<>(null, current);
        }
        tail = current;
        head.next = tail;
        head = head.next;
    }

    T read() throws BufferIOException {
        T value = tail.value;
        if (value == null) throw new BufferIOException("Tried to read from empty buffer");
        tail.value = null;
        tail = tail.next;
        return value;
    }

    void write(T data) throws BufferIOException {
        if (head.value != null) throw new BufferIOException("Tried to write to full buffer");
        head.value = data;
        head = head.next;
    }

    void overwrite(T data) {
        if (head.value == null) {
            try {
                write(data);
            } catch (BufferIOException e) {
                throw new AssertionError("Unreachable");
            }
        } else {
            head.value = data;
            head = head.next;
            tail = tail.next;
        }
    }

    void clear() {
        while (tail.value != null) {
            tail.value = null;
            tail = tail.next;
        }
    }

}

class Node<T> {
    T value;
    Node<T> next;

    public Node(T value, Node<T> next) {
        this.value = value;
        this.next = next;
    }
}

File: solutions/python/circular-buffer/1/circular_buffer.py
class BufferFullException(BufferError):
    """Exception raised when CircularBuffer is full.

    message: explanation of the error.

    """
    def __init__(self, message):
        self.message = message


class BufferEmptyException(BufferError):
    """Exception raised when CircularBuffer is empty.

    message: explanation of the error.

    """
    def __init__(self, message):
        self.message = message


class CircularBuffer:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.elements = [None] * capacity
        self.read_pos = self.write_pos = 0

    def read(self):
        if not (element := self.elements[self.read_pos]):
            raise BufferEmptyException("Circular buffer is empty")
        self.elements[self.read_pos] = None
        self.read_pos = self.shift(self.read_pos)
        return element
        

    def write(self, data):
        if self.elements[self.write_pos]:
            raise BufferFullException("Circular buffer is full")
        self.elements[self.write_pos] = data
        self.write_pos = self.shift(self.write_pos)

    def overwrite(self, data):
        if self.elements[self.write_pos]:
            self.elements[self.write_pos] = data
            self.write_pos = self.shift(self.write_pos)
            self.read_pos = self.shift(self.read_pos)
        else:
            self.write(data)

    def clear(self):
        self.elements = [None] * self.capacity

    def shift(self, position):
        return (position + 1) % self.capacity
